source/docs/guide/src/nonlinear_bitvec.md

```diff
@@ -0,0 +1,287 @@
+        x <= y,
+        x % d <= y % d,
+        y - x < d
+    ensures
+        y % d - x % d == y - x
+{
+    let small_x = x % d;
+    let small_y = y % d;
+    let tmp1 = (small_y - small_x) % d;
+    let tmp2 = (y - x) % d;
+    lemma_mod_difference_equal_helper(x,y,d, small_x, small_y, tmp1, tmp2);
+}
+```
+
+In the below `lemma_mod_between`, we want to prove that `x % d <= z % d < y % d`. However, `integer_ring` only supports equalities, so we need cannot prove `lemma_mod_between` directly. Instead, we provide facts that can help assist the proof. The helper lemma provides 1) `x % d - y % d == x - y  (mod d)` and 2) ` y % d - z % d == y - z  (mod d)`. The rest of the proof is done by `by(nonlinear_arith)`.
+
+```rust
+pub proof fn lemma_mod_between_helper(x: int, y: int, d: int, small_x:int, small_y:int, tmp1:int) by(integer_ring)
+    requires
+        small_x == x % d,
+        small_y == y % d,
+        tmp1 == (small_x - small_y) % d,
+    ensures
+        (tmp1 - (x-y)) % d == 0
+{}
+
+// note that below two facts are from the helper function, and the rest are done by `by(nonlinear_arith)`.
+// x % d - y % d == x - y  (mod d)
+// y % d - z % d == y - z  (mod d)
+pub proof fn lemma_mod_between(d: int, x: int, y: int, z: int) by(nonlinear_arith)
+    requires
+        d > 0,
+        x % d < y % d,
+        y - x <= d,
+        x <= z < y
+    ensures
+        x % d <= z % d < y % d
+{
+    let small_x = x % d;
+    let small_y = y % d;
+    let small_z = z % d;
+    let tmp1 = (small_x - small_z) % d;
+    lemma_mod_between_helper(x,z,d, small_x, small_z, tmp1);
+
+    let tmp2 = (small_z - small_y) % d;
+    lemma_mod_between_helper(z,y,d, small_z, small_y, tmp2);    
+}
+```
+
+
+#### Example 2: Proving properties on bounded integers with the help of `integer_ring`
+
+Since `integer_ring` proofs only support `int`, you need to include explicit bounds when you want to prove properties about bounded integers. For example, in order to use the proof `lemma_mod_after_mul` on `u32`s, `lemma_mod_after_mul_u32` must ensure that all arguments are within the proper bounds before passing them to `lemma_mod_after_mul`.  
+
+If a necessary bound (e.g., `m > 0`) is not included, Verus will fail to verify the proof.
+
+```rust
+proof fn lemma_mod_after_mul(x: int, y: int, z: int, m: int) by (integer_ring)
+    requires (x-y) % m == 0
+    ensures (x*z - y*z) % m == 0
+{}
+
+proof fn lemma_mod_after_mul_u32(x: u32, y: u32 , z: u32, m: u32)   
+    requires
+        m > 0,
+        (x-y) % (m as int) == 0,
+        x >= y,
+        x <= 0xffff,
+        y <= 0xffff,
+        z <= 0xffff,
+        m <= 0xffff,
+    ensures (x*z - y*z) % (m as int) == 0
+{ 
+  lemma_mod_after_mul(x as int, y as int, z as int, m as int);
+  // rest of proof body omitted for space
+}
+```
+
+The desired property for `nat` can be proved similarly. Note that we introduce several additional variables(`ab`, `bc`, and `abc`) to help with the `integer_ring` proof. 
```

Comment:
If I'm interpreting this correctly, this paragraph should be split into one piece that finishes the discussion of the example above, and a new piece that introduces the next example.   E.g.,:

> The desired property for `nat` can be proved similarly.
>
> The next example is similar, but note that we introduce several additional variables(`ab`, `bc`, and `abc`) to help with the `integer_ring` proof.

Reply:
It's unclear to me why the variable introductions are needed in the example.  The advice earlier was to introduce additional variables to deal with division.